cirq/sim/density_matrix_simulator_test.py

```diff
@@ -115,10 +115,30 @@ def test_run_channel(dtype):
     assert q0_measurements == {0, 1}
 
 
+class CountingSimulator(cirq.DensityMatrixSimulator):
+    """
+    CountingSimulator is a test class that decorates the _base_iterator method
+    so that we can count the number of circuit runs during a test case
+    """
+    def __init__(self, dtype=np.complex64):
+        super().__init__(dtype)
+        self.count = 0
+
+    def _base_iterator(
```

Comment:
I think that this kind of interception is more typically done via mocking.

Reply:
Of course! Thanks.
it was one of those late night moments where I looked and couldn't find it in unittest.mock and then I just implemented myself.
yup, it's much simpler with mocking library. change is coming...